cpp/arcticdb/entity/atom_key.hpp

```diff
@@ -101,10 +102,14 @@ class AtomKeyImpl {
 
     size_t get_cached_hash() const {
         if (!hash_) {
-            // arcticdb::commutative_hash_combine needs extra template specialisations for our variant types, folly's
-            // built-in variant forwards to std::hash which should be good enough for these simple types
-            hash_ = folly::hash::hash_combine(id_, version_id_, creation_ts_, content_hash_, key_type_, index_start_,
-                    index_end_);
+            auto seed = hash(id_);
+            boost::hash_combine(seed, version_id_);
+            boost::hash_combine(seed, creation_ts_);
+            boost::hash_combine(seed, content_hash_);
+            boost::hash_combine(seed, key_type_);
+            boost::hash_combine(seed, index_start_);
+            boost::hash_combine(seed, index_end_);
+            hash_ = seed;
```

Comment:
I think the returned values of `get_cached_hash` must not change for backward compatibility (_e.g._ to be able to retrieve segments which have been stored in databases in previous versions).

Is there a way we could test that?

Reply:
I don't think that's the case, however I agree we should check. I believe we have version backwards and forwards compatibility checks, I'll find out where they are and point you at them